ITT we get lazy

RemieRichards opened the pull request:
* ```LAZYREMOVE(L, I)``` Removes item I from list L, and then checks to see if L should be emptied using the same logic as ```UNSETEMPTY(L)```
* ```LAZYADD(L, I)``` Checks if list L is null, and initialises it if so, then adds item I to list L
* ```LAZYACCESS(L, I)``` Returns the item at index I in list L, or null

RemieRichards committed:
Adds a LAZYREMOVE macro

LAZYREMOVE(L, I) removes item I from list L, and then checks to see if L should be emptied using UNSETEMPTY(L)

RemieRichards commented:
sure, gimme 2 ticks.

MrStonedOne commented:
2 ticks are over, i don't see it.

RemieRichards committed:
Adds a LAZYADD() macro

Checks if list L is null, and initialises it if so, then adds item I to list L

RemieRichards commented:
@MrStonedOne what are your thoughts on:

```
#define LAZYINDEX(L, index) (L ? ((IsInteger(index) && (index > 0 && index <= L.len)) ? L[index] : (!isnum(index) ? L[index] : FALSE)) : FALSE)
```
and
```
#define LAZYVAR(D, V) (D ? D.vars[V] : FALSE )
```

YotaXP commented:
This checks if `L`'s trueness twice.  For something this simple, I'd manually inline the `UNSETEMPTY`

RemieRichards commented:
Yeah sure was going to do that.

Iamgoofball commented:
can we add this
`#define LAZYREMIE(G, PR) (PR.unneeded_return ? PR.closed = TRUE : PR.closed = FALSE`

RemieRichards committed:
Inlines UNSETEMPTY part of LAZYREMOVE()

For extra speed.

RemieRichards commented:
@iamgoofball I did actually write a small design doc for a tool to do just that (find trailing returns) but never built it.

YotaXP commented:
@RemieRichards Regarding `LAZYINDEX` and `LAZYVAR`, I don't think `0` is a good default return, as it may frequently be a valid value.  Why not `null`, or a third argument to the macro?

As for `LAZYINDEX`, I think non-integer numbers are acceptable for indexing.  They just get floored or rounded or something.  I seem to remember using that trick somewhere.

RemieRichards commented:
Null is probably better yes.
also yes they are floored, that's gross.

RemieRichards committed:
Adds LAZYACCESS() Macro

LAZYACCESS(L, I) returns the item at index I in list L, or null

RemieRichards commented:
Added the new ```LAZYINDEX()``` as ```LAZYACCESS()``` since it actually already does ```LAZYVAR()```'s job too.

Iamgoofball commented:
can we add LAZYCODE() while we're at it so i dont have to make any more PRs?

RemieRichards commented:
that's like, the future of programming, isn't it? when people describe what kind of programming it is, like how it used to be much more hands on with assembly, and before that machine code, and like how these days there are visual languages, p.much the apparent "goal" is to replace programming with telling a machine what you want it to do in what amounts to just normal conversation.

which I think personally takes the fun out of it but hey ~science~

YotaXP commented on code/__HELPERS/lists.dm:
```diff
@@ -424,5 +424,7 @@
-
+#define LAZYREMOVE(L, I) if(L) { L -= I; if(!L.len) { L = null; } }
+#define LAZYADD(L, I) if(!L) { L = list(); } L += I;
+#define LAZYACCESS(L, I) (L ? (isnum(I) ? (I > 0 && I <= L.len ? L[I] : null) : L[I]) : null)
```
If I were to be picky, this should be `(I >= 1 && I < L.len + 1`

optimumtact commented on code/__HELPERS/lists.dm:
```diff
@@ -424,5 +424,7 @@
+#define LAZYREMOVE(L, I) if(L) { L -= I; if(!L.len) { L = null; } }
+#define LAZYADD(L, I) if(!L) { L = list(); } L += I;
+#define LAZYACCESS(L, I) (L ? (isnum(I) ? (I > 0 && I <= L.len ? L[I] : null) : L[I]) : null)
 #define LAZYLEN(L) ( L ? L.len : 0 )
```
12:31 <%oranges> !dm var/list/x = null; world.log << length(x);;
12:31 < Bot32> oranges: 0

optimumtact replied:
this is probably slower due to length being a proccall nevermind

RemieRichards replied:
aww, I arrived too late to tell you that ;_;
;)

optimumtact replied:
MSO did some testing
http://pastebin.com/tA0MGUr3

apparently length is faster

RemieRichards committed:
Speeds up LAZYLEN macro

Replaces it's internals with a proccall to length(), this is quicker according to MSO.

Shadowlight213 commented:
@RemieRichards could you add some documentation for the lazy list stuff.

RemieRichards commented:
What's to document? they're really simple sections of code.

Shadowlight213 commented:
True.
I think it's more along the lines of we need to put down that these defines and stuff exist somewhere more obvious, along with when they should and shouldn't be used.
Mainly so people don't first learn that these exist by you telling them to use it in their PRs :P
